### PR TITLE
[0.17, WIP] Confidential Assets

### PR DESCRIPTION
**Pull request by stevenroose**
TODO:

Multi-change output script stuff
Squash history

### CODE REVIEW

**Comment by instagibbs** [comment]
All unit tests now working, including blind_tests and additional bitcoin-tx cases for elements serialization

**Comment by instagibbs** [comment]
All bitcoin functional tests are working locally for me now

File: src/blech32.cpp [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+uint32_t PolyMod(const data& v)
+{
+    // The input is interpreted as a list of coefficients of a polynomial over F = GF(32), with an
+    // implicit 1 in front. If the input is [v0,v1,v2,v3,v4], that polynomial is v(x) =
+    // 1*x^5 + v0*x^4 + v1*x^3 + v2*x^2 + v3*x + v4. The implicit 1 guarantees that
+    // [v0,v1,v2,...] has a distinct checksum from [0,v0,v1,v2,...].
+
+    // The output is a 30-bit integer whose 5-bit groups are the coefficients of the remainder of
+    // v(x) mod g(x), where g(x) is the Blech32 generator,
+    // x^6 + {29}x^5 + {22}x^4 + {20}x^3 + {21}x^2 + {29}x + {18}. g(x) is chosen in such a way
```
Comment (sipa) [inline src/blech32.cpp]: The correct Blech32 generator is
```
x^12 + {31}x^10 + {10}x^9 + {18}x^8 + {31}x^7 + {14}x^6 + {18}x^5 + {23}x^3 + {22}x^2 + {4}x + {6}
```

File: src/blech32.cpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    // that the resulting code is a BCH code, guaranteeing detection of up to 3 errors within a
```
Comment (sipa) [inline src/blech32.cpp]: Guarantees detection of up to 6 errors within a window of 1023 characters. The next sentence can be dropped (it wasn't selected for anything in particular).

File: src/blech32.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    uint64_t c = 1;
+    for (const auto v_i : v) {
+        // We want to update `c` to correspond to a polynomial with one extra term. If the initial
+        // value of `c` consists of the coefficients of c(x) = f(x) mod g(x), we modify it to
+        // correspond to c'(x) = (f(x) * x + v_i) mod g(x), where v_i is the next input to
+        // process. Simplifying:
+        // c'(x) = (f(x) * x + v_i) mod g(x)
+        //         ((f(x) mod g(x)) * x + v_i) mod g(x)
+        //         (c(x) * x + v_i) mod g(x)
+        // If c(x) = c0*x^5 + c1*x^4 + c2*x^3 + c3*x^2 + c4*x + c5, we want to compute
```
Comment (sipa) [inline src/blech32.cpp]: These get 12 terms instead of 6.

File: src/blech32.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        // c'(x) = (c0*x^5 + c1*x^4 + c2*x^3 + c3*x^2 + c4*x + c5) * x + v_i mod g(x)
+        //       = c0*x^6 + c1*x^5 + c2*x^4 + c3*x^3 + c4*x^2 + c5*x + v_i mod g(x)
+        //       = c0*(x^6 mod g(x)) + c1*x^5 + c2*x^4 + c3*x^3 + c4*x^2 + c5*x + v_i
+        // If we call (x^6 mod g(x)) = k(x), this can be written as
+        // c'(x) = (c1*x^5 + c2*x^4 + c3*x^3 + c4*x^2 + c5*x + v_i) + c0*k(x)
+
+        // First, determine the value of c0:
+        uint8_t c0 = c >> 55; // ELEMENTS: 25->55
+
+        // Then compute c1*x^5 + c2*x^4 + c3*x^3 + c4*x^2 + c5*x + v_i:
```
Comment (sipa) [inline src/blech32.cpp]: C  values from 1 though 11.

**Comment by instagibbs** [comment]
Seems to have previous branch +- `s/elementswitness/elementsmode/`, some test renaming, and debugging printing being removed.

**Comment by stevenroose** [comment]
Closing in favor of #515.

**Pull request closed without merging**